snooty/gizaparser/domain.py

```diff
@@ -0,0 +1,210 @@
+                giza_category.add(fileid, text, artifacts, entry[2] + diagnostics)
+
+            yield from self.generate_pages(prefix, all_diagnostics)
+
+    def categorize(self) -> Dict[str, List[n.FileId]]:
+        """Scan the source directory for YAML files we should ingest, and categorize them."""
+        # Categorize our YAML files
+        logger.debug("Categorizing YAML files")
+        categorized: Dict[str, List[n.FileId]] = collections.defaultdict(list)
+        for path in self.config.get_files_by_extension((".yaml",)):
+            prefix = get_giza_category(path)
+            if prefix in self.yaml_mapping:
+                categorized[prefix].append(path)
+
+        return categorized
+
+    def generate_pages(
+        self, category_name: str, all_diagnostics: Dict[n.FileId, List[Diagnostic]]
+    ) -> Iterable[Tuple[Page, Sequence[Diagnostic]]]:
+        """Generate a Page for each node in each of our managed categories."""
+        # Now that all of our YAML files are loaded, generate a page for each one
+        giza_category = self.yaml_mapping[category_name]
+        logger.debug("Processing %s YAML: %d nodes", category_name, len(giza_category))
+        for file_id, giza_node in giza_category.reify_all_files(all_diagnostics):
+
+            def create_page(filename: str) -> Tuple[Page, EmbeddedRstParser]:
+                page = Page.create(
+                    giza_node.path,
+                    filename,
+                    giza_node.text,
+                    n.Root((-1,), [], giza_node.path, {}),
+                )
+                return (
+                    page,
+                    self.rst_parser_factory(
+                        self.config, page, giza_node.parse_diagnostics
+                    ),
+                )
+
+            for page in giza_category.to_pages(giza_node.path, create_page, giza_node):
+                assert giza_node.pages is not None
+                yield (page, giza_node.diagnostics)
+
+    def update(
+        self,
+        path: n.FileId,
+        optional_text: Optional[str] = None,
+    ) -> Iterable[Tuple[Page, Sequence[Diagnostic]]]:
+        """Update an individual Giza file."""
+        file_id = path.name
+        prefix = get_giza_category(path)
+        giza_category = self.yaml_mapping[prefix]
+        needs_rebuild = set((file_id,)).union(
+            *(
+                category.dg.predecessors(file_id)
+                for category in self.yaml_mapping.values()
+            )
+        )
+        logger.debug("needs_rebuild: %s", ",".join(needs_rebuild))
+        for file_id in needs_rebuild:
+            file_diagnostics: List[Diagnostic] = []
+            try:
+                giza_node = giza_category.reify_file_id(file_id)
+            except KeyError:
+                logging.warn("No file found in registry: %s", file_id)
+                continue
+
+            steps, text, parse_diagnostics = giza_category.parse(path, optional_text)
```

Comment:
Question: What is `optional_text` exactly? Is it the text data we want to include in the page? 

If my understanding is correct, every node that needs to be rebuilt will have the same `text` information in the page node that is created for it. Is this intended?

Reply:
We can either obtain file text from the filesystem or from the language server: `optional_text` allows us to manually specify the text of a given file, so that changes don't have to be saved to disk to be parsed!

Reply:
Gotcha! That makes sense to me. 

I feel like I may be mistaken, but it appears that if we provide `optional_text` here, then all of the nodes that need to be rebuilt will have the same `text` value provided.